Source/WebGPU/WebGPU/RenderPassEncoder.mm

```diff
@@ -115,26 +140,80 @@
 
 void RenderPassEncoder::endOcclusionQuery()
 {
+    if (!prepareTheEncoderState())
+        return;
+
+    if (!m_occlusionQueryActive) {
+        makeInvalid();
+        return;
+    }
+
+    m_occlusionQueryActive = false;
     [m_renderCommandEncoder setVisibilityResultMode:MTLVisibilityResultModeDisabled offset:m_visibilityResultBufferOffset];
 }
 
 void RenderPassEncoder::endPass()
 {
+    auto& parentEncoder = m_commandEncoder.get();
+    if (state() != CommandsMixin::EncoderState::Open || parentEncoder.state() != CommandsMixin::EncoderState::Locked) {
+        m_device->generateAValidationError("endPass failed valiation"_s);
+        return;
+    }
+
+    setState(CommandsMixin::EncoderState::Ended);
+    parentEncoder.setState(CommandsMixin::EncoderState::Open);
+
+    if (!isValid() || m_occlusionQueryActive || m_debugGroupStackSize) {
+        parentEncoder.makeInvalid();
+        return;
+    }
+
     ASSERT(m_pendingTimestampWrites.isEmpty() || m_device->baseCapabilities().counterSamplingAPI == HardwareCapabilities::BaseCapabilities::CounterSamplingAPI::CommandBoundary);
     for (const auto& pendingTimestampWrite : m_pendingTimestampWrites)
         [m_renderCommandEncoder sampleCountersInBuffer:pendingTimestampWrite.querySet->counterSampleBuffer() atSampleIndex:pendingTimestampWrite.queryIndex withBarrier:NO];
     m_pendingTimestampWrites.clear();
     [m_renderCommandEncoder endEncoding];
     m_renderCommandEncoder = nil;
+    m_pipelineLayout = nullptr;
 }
 
 void RenderPassEncoder::endPipelineStatisticsQuery()
 {
 
 }
 
+static bool validToUseWith(auto& object, auto& targetObject)
```

Comment:
We already have `IsValidToUseWith.h`

Reply:
Oh great I’ll use that instead, thank you!